File: integrations/hermes-airhop-channel-gateway/src/airhop_hermes_gateway/whatsapp_templates.py
"""Restricted, inspectable Meta utility templates for staff-approved replies."""

from __future__ import annotations
import re
from typing import Any


def normalize_template(value: Any) -> dict | None:
    if (
        not isinstance(value, dict)
        or value.get("status") != "APPROVED"
        or value.get("category") != "UTILITY"
    ):
        return None
    name, language = value.get("name"), value.get("language")
    if (
        not isinstance(name, str)
        or not re.fullmatch(r"[a-z0-9_]{1,512}", name)
        or not isinstance(language, str)
        or not re.fullmatch(r"[a-z]{2,3}(?:_[A-Z]{2})?", language)
    ):
        return None
    parts = value.get("components")
    if not isinstance(parts, list) or not 1 <= len(parts) <= 3:
        return None
    texts = {}
    for part in parts:
        if not isinstance(part, dict):
            return None
        kind, text = part.get("type"), part.get("text")
        if (
            kind not in {"BODY", "HEADER", "FOOTER"}
            or kind in texts
            or not isinstance(text, str)
            or not text.strip()
            or len(text) > 4096
        ):
            return None
        if kind == "HEADER" and part.get("format") != "TEXT":
            return None
        if kind != "BODY" and ("{" in text or "}" in text):
            return None
        texts[kind] = text
    if "BODY" not in texts:
        return None
    body = texts["BODY"]
    slots = {int(i) for i in re.findall(r"\{\{([1-9][0-9]?)\}\}", body)}
    if (
        len(slots) > 10
        or slots != set(range(1, len(slots) + 1))
        or re.search(r"[{}]", re.sub(r"\{\{[1-9][0-9]?\}\}", "", body))
    ):
        return None
    return {
        "name": name,
        "language": language,
        "body": body,
        "header": texts.get("HEADER", ""),
        "footer": texts.get("FOOTER", ""),
        "parameterCount": len(slots),
    }


def render_template(template: dict, parameters: Any) -> str:
    if (
        not isinstance(parameters, list)
        or len(parameters) != template["parameterCount"]
        or any(
            not isinstance(v, str)
            or not v.strip()
            or len(v) > 500
            or any(c in v for c in "\n\r\t{}")
            for v in parameters
        )
    ):
        raise ValueError("invalid WhatsApp template parameters")
    body = re.sub(
        r"\{\{([1-9][0-9]?)\}\}", lambda m: parameters[int(m[1]) - 1], template["body"]
    )
    rendered = "\n".join(
        part for part in (template["header"], body, template["footer"]) if part
    )
    if len(rendered) > 4096:
        raise ValueError("WhatsApp template message too long")
    return rendered

File: crates/buzz-db/src/airhop/external_conversation/whatsapp_templates.rs
//! Staff-reviewed template messages use the ordinary signed event/outbox path.
use super::*;

pub(super) async fn validate_template(
    tx: &mut sqlx::Transaction<'_, sqlx::Postgres>,
    community_id: Uuid,
    connection_id: Uuid,
    event: &Event,
    actor_kind: &str,
) -> Result<()> {
    let tags: Vec<_> = event
        .tags
        .iter()
        .filter(|t| {
            t.as_slice()
                .first()
                .is_some_and(|v| v == "airhop-whatsapp-template")
        })
        .collect();
    if tags.is_empty() {
        return Ok(());
    }
    if actor_kind != "staff" || tags.len() != 1 || tags[0].as_slice().len() != 2 {
        return Err(DbError::AccessDenied(
            "WhatsApp templates require an explicit staff message".into(),
        ));
    }
    #[derive(serde::Deserialize)]
    #[serde(deny_unknown_fields)]
    struct Spec {
        name: String,
        language: String,
        parameters: Vec<String>,
    }
    let spec: Spec = serde_json::from_str(&tags[0].as_slice()[1])?;
    if spec.parameters.len() > 10
        || spec.parameters.iter().any(|v| {
            v.trim().is_empty()
                || v.chars().count() > 500
                || v.contains(['\n', '\r', '\t', '{', '}'])
        })
    {
        return Err(DbError::InvalidData(
            "invalid WhatsApp template parameters".into(),
        ));
    }
    // Serialize sends per connection for the hourly bound. Provider catalog is
    // supplied by the authenticated gateway and rechecked with Meta at dispatch.
    sqlx::query("SELECT pg_advisory_xact_lock(hashtextextended($1,0))")
        .bind(format!("whatsapp-templates:{community_id}:{connection_id}"))
        .execute(&mut **tx)
        .await?;
    let caps:Option<Value>=sqlx::query_scalar("SELECT observed_capabilities FROM airhop_channel_connections WHERE community_id=$1 AND id=$2 AND provider='whatsapp_cloud' AND status='active'")
        .bind(community_id).bind(connection_id).fetch_optional(&mut **tx).await?;
    let caps =
        caps.ok_or_else(|| DbError::AccessDenied("active WhatsApp connection required".into()))?;
    let synced = caps
        .get("templatesSyncedAt")
        .and_then(Value::as_i64)
        .unwrap_or(0);
    if chrono::Utc::now().timestamp() - synced > 600
        || synced > chrono::Utc::now().timestamp() + 300
    {
        return Err(DbError::InvalidData(
            "WhatsApp templates need synchronization".into(),
        ));
    }
    let template = caps
        .get("utilityTemplates")
        .and_then(Value::as_array)
        .and_then(|items| {
            items
                .iter()
                .find(|t| t["name"] == spec.name && t["language"] == spec.language)
        })
        .ok_or_else(|| {
            DbError::InvalidData("approved WhatsApp utility template unavailable".into())
        })?;
    let mut body = template
        .get("body")
        .and_then(Value::as_str)
        .ok_or_else(|| DbError::InvalidData("invalid WhatsApp template body".into()))?
        .to_owned();
    if template.get("parameterCount").and_then(Value::as_u64) != Some(spec.parameters.len() as u64)
    {
        return Err(DbError::InvalidData(
            "WhatsApp template parameters changed".into(),
        ));
    }
    for (i, value) in spec.parameters.iter().enumerate() {
        body = body.replace(&format!("{{{{{}}}}}", i + 1), value);
    }
    let rendered = [
        template.get("header").and_then(Value::as_str).unwrap_or(""),
        body.as_str(),
        template.get("footer").and_then(Value::as_str).unwrap_or(""),
    ]
    .into_iter()
    .filter(|v| !v.is_empty())
    .collect::<Vec<_>>()
    .join("\n");
    if rendered != event.content || rendered.chars().count() > 4096 || rendered.contains(['{', '}'])
    {
        return Err(DbError::InvalidData(
            "WhatsApp template preview no longer matches approved content".into(),
        ));
    }
    let duplicate:bool=sqlx::query_scalar("SELECT EXISTS(SELECT 1 FROM airhop_external_message_outbox WHERE community_id=$1 AND buzz_event_id=$2)")
        .bind(community_id).bind(event.id.as_bytes().as_slice()).fetch_one(&mut **tx).await?;
    if !duplicate {
        let count:i64=sqlx::query_scalar("SELECT count(*) FROM airhop_external_message_outbox WHERE community_id=$1 AND connection_id=$2 AND created_at>now()-interval '1 hour' AND EXISTS(SELECT 1 FROM jsonb_array_elements(event_json->'tags') tag WHERE tag->>0='airhop-whatsapp-template')")
            .bind(community_id).bind(connection_id).fetch_one(&mut **tx).await?;
        if count >= 20 {
            return Err(DbError::InvalidData(
                "WhatsApp service template hourly limit reached".into(),
            ));
        }
    }
    Ok(())
}
